### ISSUE TITLE
Extract whole archives in some cases

### ISSUE THREAD

**Issue by fr500** [libretro/RetroArch#4331]
Something like this may be desirable for some file types:
https://hastebin.com/eburoyuzaq.cs

Basically there should be a list of extensions that trigger a whole archive extraction instead of just the selected file. It would be slow but it would solve a bunch of cases with the current code

I tried to implement it but the extraction function is quite complicated.

**Comment by fr500** [comment]
https://github.com/libretro/dosbox-libretro/issues/49

**Comment by fr500** [comment]
https://github.com/libretro/RetroArch/issues/4094

**Comment by fr500** [comment]
https://github.com/libretro/RetroArch/issues/2614

**Comment by RobLoach** [comment]
File extensions that would likely need this:

```
bat com exe game cue gdi toc wad m3u
```

**Comment by klepp0906** [comment]
Just chiming in hoping this is on the menu to be addressed. Otherwise I have to literally convert an entire Psx library. 

I posted this here thinking it was a core issue considering Sega cd bin/ cue seem to extract and load fine. Psx not so much. 

https://github.com/libretro/beetle-psx-libretro/issues/223

**Comment by SpaceAgeHero** [comment]
I've posted a bounty on this. :-)
Please kindly also take into account my referenced request https://github.com/libretro/RetroArch/issues/2614 which includes **multi-bin dumps as well**.

**Comment by Ferk** [comment]
I honestly think that it would be a bad (and hopefully temporary) solution to simply unpack all the files in the zip.

Not all platforms have that kind of readily available storage, and imho it defeats the point of zipping the games if each game I want to play is gonna be unzipped into disk. And even if they were deleted afterwards it's a lot of strain for the storage, specially if you think about SD cards with Lakka and so.

Best would be if an VFS (#3715?) could read the files directly from the compressed file without dumping them to disk. That's how I initially thought the zip support in Retroarch worked, I was a bit disappointed when I discovered it's just unzipping them to disk.

**Comment by bparker06** [comment]
@Ferk What about zips with e.g. multi-disc games? People keep asking for this but I think it's clear we don't want to keep all that in RAM.

**Comment by klepp0906** [comment]
I just keep my seperate discs in seperate archives. I realize not everyone does it this way however. 

Unzipping all to a temp directory same as everything else is the best solution I reckon.

**Comment by RobLoach** [comment]
In order to save RAM, disk space, and performance, extracting the whole archive should be opt-in by the user. It currently extracts the first file in the archive. The code is over at:
https://github.com/libretro/RetroArch/blob/master/tasks/task_content.c#L454

**Comment by Ferk** [comment]
@bparker06 I don't think you need to extract the full file into RAM if you only want to read part of it. Not with the typical deflate algorithm used in zip files, and I expect the lzma from 7z would also allow uncompressing files partially ([checking at the lib7z code I can see there are read and seek functions](https://github.com/sina-ht/lib7z/blob/master/include/7z.h#L96-L97)).

I assume a VFS would be able to uncompress on the fly big files as they are read, without keeping the entire thing in RAM. It might be slower than caching them on disk, but it probably will still be faster than the physical optical devices from actual consoles.

**Comment by SpaceAgeHero** [comment]
Any clue why my bounty https://www.bountysource.com/issues/40462238-extract-whole-archives-in-some-cases is not referenced? Wish this would get more attention...

**Comment by klepp0906** [comment]
I haven’t forgotten I tell ya that much. This is a huge deal. Amazed it’s still an issue this many years in. 

Ty for the bounty btw <3

**Comment by Tatsuya79** [comment]
It would be really nice if it could already unzip everything to disk.

Use case: cores that read games from multiple floppies and have a GUI to load them (NP2Kai, PX68K).
If you load a zip that gets entirely decompressed into a cache folder, then you just need those cores config path to point to that same folder to have the exact files you need usable in their GUI.
That's something really neat as those GUI are particularly awful for browsing a huge amount of files.
That works for MAME Software List sets as they are using multiple disks in a single archive.

I wonder if that simply could be an option in retroarch settings... core? 
"Extract everything from loaded archive." and if that could be used as a core override as well.

**Comment by Ferk** [comment]
Wouldn't the VFS pull (#5664) together with PhysFS integration (RobLoach#3) allow any core supporting the VFS to access all the content of the zip files without having to extract anything at all to disk?

**Comment by i30817** [comment]
>  In order to save RAM, disk space, and performance, extracting the whole archive should be opt-in by the user.

Extracting should be already optional (it would be the first thing i'd disable). But isn't for some reason.
